<p>The Data used for below analysis has been provided by <a href="https://www.worldbank.org">The World Bank</a>. </p>
<p>Exploring and analyzing international debt data collected by The World Bank. The dataset contains information about the amount of debt (in USD) owed by developing countries across several categories (debt indicators) w.r.t the year <b>2020</b>.

In [1]:
%load_ext sql

In [2]:
from sqlalchemy import create_engine

In [11]:
%sql postgresql://postgres:root1234@localhost/IDS_analysis

'Connected: postgres@IDS_analysis'

## 1. The World Bank's international debt data


In [12]:
%%sql

SELECT *
FROM debt
WHERE debt_amount IS NOT NULL
LIMIT 10;

 * postgresql://postgres:***@localhost/IDS_analysis
10 rows affected.


[('Afghanistan', 'AFG', 'DT.TDS.DEGG.CD', 'Debt service on external debt, general government sector (PPG) (TDS, current US$)', 127503239.8),
 ('Afghanistan', 'AFG', 'DT.TDS.DLXF.CD', 'Debt service on external debt, long-term (TDS, current US$)', 134184969.8),
 ('Afghanistan', 'AFG', 'DT.TDS.DPNG.CD', 'Debt service on external debt, private nonguaranteed (PNG) (TDS, current US$)', 6681730.0),
 ('Afghanistan', 'AFG', 'DT.TDS.DPPG.CD', 'Debt service on external debt, public and publicly guaranteed (PPG) (TDS, current US$)', 127503239.8),
 ('Afghanistan', 'AFG', 'DT.TDS.DEPS.CD', 'Debt service on external debt, public sector (PPG) (TDS, current US$)', 127503239.8),
 ('Afghanistan', 'AFG', 'DT.TDS.DECT.CD', 'Debt service on external debt, total (TDS, current US$)', 134184969.8),
 ('Afghanistan', 'AFG', 'DT.DIS.DLXF.CD', 'Disbursements on external debt, long-term (DIS, current US$)', 85548445.6),
 ('Afghanistan', 'AFG', 'DT.DIS.DPPG.CD', 'Disbursements on external debt, public and publicly guaranteed (PPG) (DIS, current US$)', 85548445.6),
 ('Afghanistan', 'AFG', 'DT.AMT.BLAT.GG.CD', 'GG, bilateral (AMT, current US$)', 47734121.8),
 ('Afghanistan', 'AFG', 'DT.INT.BLAT.GG.CD', 'GG, bilateral (INT, current US$)', 27913192.7)]

## 2. Finding the number of distinct countries

In [13]:
%%sql

SELECT 
    COUNT(DISTINCT(country_name)) AS total_distinct_countries
FROM debt
WHERE debt_amount IS NOT NULL;

 * postgresql://postgres:***@localhost/IDS_analysis
1 rows affected.


[(120,)]

## 3. Finding out the distinct debt indicators

In [15]:
%%sql

SELECT 
    DISTINCT(indicator_code) as distinct_debt_indicators
FROM debt
WHERE debt_amount IS NOT NULL
GROUP BY distinct_debt_indicators
ORDER BY distinct_debt_indicators;


 * postgresql://postgres:***@localhost/IDS_analysis
141 rows affected.


[('DT.AMT.BLAT.CD',),
 ('DT.AMT.BLAT.GG.CD',),
 ('DT.AMT.BLAT.OPS.CD',),
 ('DT.AMT.BLAT.PRVG.CD',),
 ('DT.AMT.BLAT.PS.CD',),
 ('DT.AMT.BLTC.CD',),
 ('DT.AMT.DEGG.CD',),
 ('DT.AMT.DLXF.CD',),
 ('DT.AMT.DOPS.CD',),
 ('DT.AMT.DPNG.CD',),
 ('DT.AMT.DPPG.CD',),
 ('DT.AMT.MLAT.CD',),
 ('DT.AMT.MLAT.GG.CD',),
 ('DT.AMT.MLAT.OPS.CD',),
 ('DT.AMT.MLAT.PRVG.CD',),
 ('DT.AMT.MLAT.PS.CD',),
 ('DT.AMT.MLTC.OPS.CD',),
 ('DT.AMT.MLTC.PS.CD',),
 ('DT.AMT.OFFT.CD',),
 ('DT.AMT.OFFT.GG.CD',),
 ('DT.AMT.OFFT.OPS.CD',),
 ('DT.AMT.OFFT.PRVG.CD',),
 ('DT.AMT.OFFT.PS.CD',),
 ('DT.AMT.PBND.CD',),
 ('DT.AMT.PBND.GG.CD',),
 ('DT.AMT.PBND.OPS.CD',),
 ('DT.AMT.PBND.PS.CD',),
 ('DT.AMT.PCBK.CD',),
 ('DT.AMT.PCBK.GG.CD',),
 ('DT.AMT.PCBK.OPS.CD',),
 ('DT.AMT.PCBK.PRVG.CD',),
 ('DT.AMT.PCBK.PS.CD',),
 ('DT.AMT.PNGB.CD',),
 ('DT.AMT.PNGC.CD',),
 ('DT.AMT.PROP.CD',),
 ('DT.AMT.PROP.GG.CD',),
 ('DT.AMT.PROP.OPS.CD',),
 ('DT.AMT.PROP.PS.CD',),
 ('DT.AMT.PRPG.CD',),
 ('DT.AMT.PRVT.CD',),
 ('DT.AMT.PRVT.GG.CD',),
 ('DT.AMT.PRVT.OPS.CD',),
 ('DT.AMT.PRVT.PRVG.CD',),
 ('DT.AMT.PRVT.PS.CD',),
 ('DT.DIS.BLAT.CD',),
 ('DT.DIS.DLXF.CD',),
 ('DT.DIS.DPPG.CD',),
 ('DT.DIS.MLAT.CD',),
 ('DT.DIS.OFFT.CD',),
 ('DT.DIS.PCBK.CD',),
 ('DT.DIS.PROP.CD',),
 ('DT.DIS.PRVT.CD',),
 ('DT.INT.BLAT.CD',),
 ('DT.INT.BLAT.GG.CD',),
 ('DT.INT.BLAT.OPS.CD',),
 ('DT.INT.BLAT.PRVG.CD',),
 ('DT.INT.BLAT.PS.CD',),
 ('DT.INT.BLTC.CD',),
 ('DT.INT.DEGG.CD',),
 ('DT.INT.DEPS.CD',),
 ('DT.INT.DLXF.CD',),
 ('DT.INT.DOPS.CD',),
 ('DT.INT.DPNG.CD',),
 ('DT.INT.DPPG.CD',),
 ('DT.INT.MLAT.CD',),
 ('DT.INT.MLAT.GG.CD',),
 ('DT.INT.MLAT.OPS.CD',),
 ('DT.INT.MLAT.PRVG.CD',),
 ('DT.INT.MLAT.PS.CD',),
 ('DT.INT.MLTC.OPS.CD',),
 ('DT.INT.MLTC.PS.CD',),
 ('DT.INT.OFFT.CD',),
 ('DT.INT.OFFT.GG.CD',),
 ('DT.INT.OFFT.OPS.CD',),
 ('DT.INT.OFFT.PRVG.CD',),
 ('DT.INT.OFFT.PS.CD',),
 ('DT.INT.PBND.CD',),
 ('DT.INT.PBND.GG.CD',),
 ('DT.INT.PBND.OPS.CD',),
 ('DT.INT.PBND.PRVG.CD',),
 ('DT.INT.PBND.PS.CD',),
 ('DT.INT.PCBK.CD',),
 ('DT.INT.PCBK.GG.CD',),
 ('DT.INT.PCBK.OPS.CD',),
 ('DT.INT.PCBK.PRVG.CD',),
 ('DT.INT.PCBK.PS.CD',),
 ('DT.INT.PNGB.CD',),
 ('DT.INT.PNGC.CD',),
 ('DT.INT.PROP.CD',),
 ('DT.INT.PROP.GG.CD',),
 ('DT.INT.PROP.OPS.CD',),
 ('DT.INT.PROP.PS.CD',),
 ('DT.INT.PRPG.CD',),
 ('DT.INT.PRVT.CD',),
 ('DT.INT.PRVT.GG.CD',),
 ('DT.INT.PRVT.OPS.CD',),
 ('DT.INT.PRVT.PRVG.CD',),
 ('DT.INT.PRVT.PS.CD',),
 ('DT.TDS.BLAT.CD',),
 ('DT.TDS.BLAT.GG.CD',),
 ('DT.TDS.BLAT.OPS.CD',),
 ('DT.TDS.BLAT.PRVG.CD',),
 ('DT.TDS.BLAT.PS.CD',),
 ('DT.TDS.BLTC.CD',),
 ('DT.TDS.DECT.CD',),
 ('DT.TDS.DEGG.CD',),
 ('DT.TDS.DEPS.CD',),
 ('DT.TDS.DLXF.CD',),
 ('DT.TDS.DOPS.CD',),
 ('DT.TDS.DPNG.CD',),
 ('DT.TDS.DPPG.CD',),
 ('DT.TDS.MLAT.CD',),
 ('DT.TDS.MLAT.GG.CD',),
 ('DT.TDS.MLAT.OPS.CD',),
 ('DT.TDS.MLAT.PRVG.CD',),
 ('DT.TDS.MLAT.PS.CD',),
 ('DT.TDS.MLTC.OPS.CD',),
 ('DT.TDS.OFFT.GG.CD',),
 ('DT.TDS.OFFT.OPS.CD',),
 ('DT.TDS.OFFT.PRVG.CD',),
 ('DT.TDS.OFFT.PS.CD',),
 ('DT.TDS.PBND.CD',),
 ('DT.TDS.PBND.GG.CD',),
 ('DT.TDS.PBND.OPS.CD',),
 ('DT.TDS.PBND.PRVG.CD',),
 ('DT.TDS.PBND.PS.CD',),
 ('DT.TDS.PCBK.CD',),
 ('DT.TDS.PCBK.GG.CD',),
 ('DT.TDS.PCBK.OPS.CD',),
 ('DT.TDS.PCBK.PRVG.CD',),
 ('DT.TDS.PCBK.PS.CD',),
 ('DT.TDS.PNGB.CD',),
 ('DT.TDS.PNGC.CD',),
 ('DT.TDS.PROP.GG.CD',),
 ('DT.TDS.PROP.OPS.CD',),
 ('DT.TDS.PROP.PS.CD',),
 ('DT.TDS.PRPG.CD',),
 ('DT.TDS.PRVT.GG.CD',),
 ('DT.TDS.PRVT.OPS.CD',),
 ('DT.TDS.PRVT.PRVG.CD',),
 ('DT.TDS.PRVT.PS.CD',)]

## 4. Total amount of debt owed by the countries in trillion USD

In [16]:
%%sql

SELECT 
    (SUM(debt_amount))/1000000000000 as total_debt
FROM debt;


 * postgresql://postgres:***@localhost/IDS_analysis
1 rows affected.


[(12.489442212632923,)]

## 5. Top 10 countries with the highest debt in billion USD

In [18]:
%%sql

SELECT 
    country_name, 
    SUM(debt_amount)/1000000000 as total_debt
FROM debt
GROUP BY country_name
ORDER BY total_debt DESC 
LIMIT 10;

 * postgresql://postgres:***@localhost/IDS_analysis
10 rows affected.


[('China', 1520.0994998066997),
 ('Indonesia', 1305.5659934938),
 ('Russian Federation', 1058.1173909981999),
 ('Brazil', 851.9726154805998),
 ('Turkey', 763.6158988021001),
 ('India', 734.4365059392001),
 ('Mexico', 620.9680577806),
 ('Argentina', 483.25932527570006),
 ('Kazakhstan', 310.6444180757),
 ('Egypt, Arab Rep.', 281.93981603020006)]

## 6. Average amount of debt across indicators in million USD

In [28]:
%%sql

SELECT 
    indicator_code AS debt_indicator,
    indicator_name,
    (AVG(debt_amount))/1000000 as average_debt
FROM debt
WHERE debt_amount IS NOT NULL
GROUP BY debt_indicator, indicator_name
ORDER BY average_debt DESC;

 * postgresql://postgres:***@localhost/IDS_analysis
141 rows affected.


[('DT.TDS.DECT.CD', 'Debt service on external debt, total (TDS, current US$)', 8554.159281270833),
 ('DT.TDS.DLXF.CD', 'Debt service on external debt, long-term (TDS, current US$)', 8494.893259384167),
 ('DT.TDS.DPNG.CD', 'Debt service on external debt, private nonguaranteed (PNG) (TDS, current US$)', 7971.45749451852),
 ('DT.TDS.PNGC.CD', 'PNG, commercial banks and other creditors (TDS, current US$)', 6811.754427700001),
 ('DT.AMT.DPNG.CD', 'Principal repayments on external debt, private nonguaranteed (PNG) (AMT, current US$)', 6702.333292560001),
 ('DT.AMT.DLXF.CD', 'Principal repayments on external debt, long-term (AMT, current US$)', 6640.538940854166),
 ('DT.AMT.PNGC.CD', 'PNG, commercial banks and other creditors (AMT, current US$)', 5852.671398960001),
 ('DT.TDS.DPPG.CD', 'Debt service on external debt, public and publicly guaranteed (PPG) (TDS, current US$)', 3114.1591034274998),
 ('DT.TDS.DEPS.CD', 'Debt service on external debt, public sector (PPG) (TDS, current US$)', 3109.5189999724994),
 ('DT.AMT.PNGB.CD', 'PNG, bonds (AMT, current US$)', 2955.3255306478263),
 ('DT.AMT.PBND.PS.CD', 'PS, bonds (AMT, current US$)', 2906.0405419549998),
 ('DT.AMT.PBND.CD', 'PPG, bonds (AMT, current US$)', 2906.0405419549998),
 ('DT.TDS.PNGB.CD', 'PNG, bonds (TDS, current US$)', 2878.494804114286),
 ('DT.TDS.PBND.CD', 'PPG, bonds (TDS, current US$)', 2707.248164850725),
 ('DT.TDS.PBND.PS.CD', 'PS, bonds (TDS, current US$)', 2706.7502161260877),
 ('DT.TDS.PRVT.PS.CD', 'PS, private creditors (TDS, current US$)', 2702.8832952336847),
 ('DT.AMT.PBND.GG.CD', 'GG, bonds (AMT, current US$)', 2466.185081110526),
 ('DT.TDS.PBND.GG.CD', 'GG, bonds (TDS, current US$)', 2185.8432603882356),
 ('DT.TDS.DEGG.CD', 'Debt service on external debt, general government sector (PPG) (TDS, current US$)', 2177.3400688399997),
 ('DT.AMT.DPPG.CD', 'Principal repayments on external debt, public and publicly guaranteed (PPG) (AMT, current US$)', 2172.319823235001),
 ('DT.AMT.PRVT.CD', 'PPG, private creditors (AMT, current US$)', 2105.2756662469133),
 ('DT.AMT.PRVT.PS.CD', 'PS, private creditors (AMT, current US$)', 2104.259813024691),
 ('DT.TDS.PRVT.OPS.CD', 'OPS, private creditors (TDS, current US$)', 2031.2090474340905),
 ('DT.AMT.PBND.OPS.CD', 'OPS, bonds (AMT, current US$)', 1877.2157163083332),
 ('DT.INT.DLXF.CD', 'Interest payments on external debt, long-term (INT, current US$)', 1854.3515571833334),
 ('DT.TDS.PRVT.GG.CD', 'GG, private creditors (TDS, current US$)', 1800.0076877010758),
 ('DT.AMT.PRVT.OPS.CD', 'OPS, private creditors (AMT, current US$)', 1715.1194937694447),
 ('DT.TDS.PBND.OPS.CD', 'OPS, bonds (TDS, current US$)', 1588.6843003375),
 ('DT.DIS.DLXF.CD', 'Disbursements on external debt, long-term (DIS, current US$)', 1569.495044324369),
 ('DT.DIS.DPPG.CD', 'Disbursements on external debt, public and publicly guaranteed (PPG) (DIS, current US$)', 1528.5457832941167),
 ('DT.AMT.DEGG.CD', 'Principal repayments on external debt, general government sector (PPG) (AMT, current US$)', 1497.2000103633343),
 ('DT.AMT.PRVT.GG.CD', 'GG, private creditors (AMT, current US$)', 1411.6979620740258),
 ('DT.INT.DPNG.CD', 'Interest payments on external debt, private nonguaranteed (PNG) (INT, current US$)', 1351.8700399728393),
 ('DT.TDS.PCBK.OPS.CD', 'OPS, commercial banks (TDS, current US$)', 1318.2162359499998),
 ('DT.DIS.OFFT.CD', 'PPG, official creditors (DIS, current US$)', 1288.0073449151253),
 ('DT.TDS.DOPS.CD', 'Debt service on external debt, other public sector (PPG) (TDS, current US$)', 1202.8115240526886),
 ('DT.AMT.PCBK.OPS.CD', 'OPS, commercial banks (AMT, current US$)', 1174.4483041250003),
 ('DT.INT.PBND.CD', 'PPG, bonds (INT, current US$)', 1022.5869811188409),
 ('DT.INT.PBND.PS.CD', 'PS, bonds (INT, current US$)', 1022.0890323942031),
 ('DT.TDS.OFFT.PS.CD', 'PS, official creditors (TDS, current US$)', 969.7363912291673),
 ('DT.INT.PNGC.CD', 'PNG, commercial banks and other creditors (INT, current US$)', 959.08302873875),
 ('DT.AMT.DOPS.CD', 'Pr

## 7. The highest amount of principal repayments in million USD

In [20]:
%%sql

SELECT 
    country_name, 
    (AVG(debt_amount))/1000000 as average_debt,
    indicator_name,
    indicator_code
FROM debt
WHERE debt_amount IS NOT NULL
GROUP BY country_name, indicator_name, indicator_code
HAVING indicator_code = 'DT.AMT.DLXF.CD'
ORDER BY average_debt DESC
LIMIT 5;

 * postgresql://postgres:***@localhost/IDS_analysis
5 rows affected.


[('China', 130526.0, 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD'),
 ('Russian Federation', 78718.034785, 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD'),
 ('Indonesia', 73741.964676, 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD'),
 ('Brazil', 68078.876466, 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD'),
 ('Turkey', 61023.491826, 'Principal repayments on external debt, long-term (AMT, current US$)', 'DT.AMT.DLXF.CD')]

## 8. Some most common debt indicators across countries

In [24]:
%%sql

SELECT 
    indicator_name,
    COUNT(indicator_code) as indicator_count
FROM debt
WHERE debt_amount IS NOT NULL
GROUP BY indicator_code, indicator_name
ORDER BY indicator_count DESC
LIMIT 35;

 * postgresql://postgres:***@localhost/IDS_analysis
35 rows affected.


[('PS, official creditors (AMT, current US$)', 120),
 ('PS, multilateral (INT, current US$)', 120),
 ('Interest payments on external debt, public sector (PPG) (INT, current US$)', 120),
 ('GG, official creditors (AMT, current US$)', 120),
 ('Debt service on external debt, public and publicly guaranteed (PPG) (TDS, current US$)', 120),
 ('Debt service on external debt, long-term (TDS, current US$)', 120),
 ('Multilateral debt service (TDS, current US$)', 120),
 ('PPG, official creditors (AMT, current US$)', 120),
 ('GG, official creditors (INT, current US$)', 120),
 ('PS, official creditors (TDS, current US$)', 120),
 ('Interest payments on external debt, general government sector (PPG) (INT, current US$)', 120),
 ('Debt service on external debt, total (TDS, current US$)', 120),
 ('PPG, official creditors (INT, current US$)', 120),
 ('GG, multilateral (TDS, current US$)', 120),
 ('GG, multilateral (INT, current US$)', 120),
 ('Principal repayments on external debt, long-term (AMT, current US$)', 120),
 ('PS, multilateral (TDS, current US$)', 120),
 ('PPG, multilateral (INT, current US$)', 120),
 ('Principal repayments on external debt, general government sector (PPG) (AMT, current US$)', 120),
 ('PS, official creditors (INT, current US$)', 120),
 ('Debt service on external debt, general government sector (PPG) (TDS, current US$)', 120),
 ('Interest payments on external debt, long-term (INT, current US$)', 120),
 ('Principal repayments on external debt, public and publicly guaranteed (PPG) (AMT, current US$)', 120),
 ('Interest payments on external debt, public and publicly guaranteed (PPG) (INT, current US$)', 120),
 ('GG, official creditors (TDS, current US$)', 120),
 ('Debt service on external debt, public sector (PPG) (TDS, current US$)', 120),
 ('PPG, bilateral (TDS, current US$)', 119),
 ('Disbursements on external debt, public and publicly guaranteed (PPG) (DIS, current US$)', 119),
 ('PS, bilateral (INT, current US$)', 119),
 ('PS, bilateral (TDS, current US$)', 119),
 ('PPG, bilateral (INT, current US$)', 119),
 ('GG, multilateral (AMT, current US$)', 119),
 ('Disbursements on external debt, long-term (DIS, current US$)', 119),
 ('PPG, official creditors (DIS, current US$)', 119),
 ('PS, multilateral (AMT, current US$)', 119)]

## 9. Maximum amount of debt across the indicators along with the respective country names in billion USD

In [25]:
%%sql

SELECT
    country_name,
    indicator_name,
    MAX(debt_amount)/1000000000 as maximum_debt
FROM debt
WHERE debt_amount IS NOT NULL
GROUP BY country_name, indicator_name
ORDER BY maximum_debt DESC
LIMIT 10;

 * postgresql://postgres:***@localhost/IDS_analysis
10 rows affected.


[('China', 'Debt service on external debt, long-term (TDS, current US$)', 159.165),
 ('China', 'Debt service on external debt, total (TDS, current US$)', 159.165),
 ('China', 'Principal repayments on external debt, long-term (AMT, current US$)', 130.526),
 ('China', 'Debt service on external debt, private nonguaranteed (PNG) (TDS, current US$)', 127.566),
 ('China', 'Principal repayments on external debt, private nonguaranteed (PNG) (AMT, current US$)', 105.184),
 ('Russian Federation', 'Debt service on external debt, long-term (TDS, current US$)', 96.787494551),
 ('Russian Federation', 'Debt service on external debt, total (TDS, current US$)', 96.787494551),
 ('Indonesia', 'Debt service on external debt, total (TDS, current US$)', 88.934581879),
 ('Indonesia', 'Debt service on external debt, long-term (TDS, current US$)', 88.934581879),
 ('Brazil', 'Debt service on external debt, total (TDS, current US$)', 85.40865742)]

## Conclusion
### Few deductions from analysis:
<ul><b>120 countries</b> have taken loan from the world bank across <b>141 debt indicators</b> which amounts to <b>12.48 trillion US dollars</b>.</ul>
<ul><b>China</b> has taken the most amount of loan (<b>1520 billion US dollars</b>) from the world bank and a  large chunk of its loan is under <b>debt service on external debt, long term(TDS, current US$)</b>.</ul>
<ul>There are a total of <b>27</b> debt indicators in which all the countries listed in our dataset have taken debt. This indicates that all these countries are suffering from some common economic issues.</ul>
